src/DynamoCoreWpf/Views/Search/LibraryView.xaml.cs

```diff
@@ -142,10 +141,41 @@ private void OnTreeViewItemPreviewMouseLeftButton(object sender, MouseButtonEven
             {
                 if (wrapPanel.MakeOrClearSelection(selectedClass))
                 {
-                    e.Handled = true;
+                    // If class button was clicked, then handle, otherwise leave it.
+                    e.Handled = selectedClass.SubCategories.Count == 0;
                     selectedElement.BringIntoView();
                 }
             }
+
+            // If class is at the same level as namespace.
+            // Close all open namespaces and select class.
+            foreach (var category in categoryButton.Items)
+            {
+                var catVM = category as NodeCategoryViewModel;
+                if (catVM == null) continue;
+
+                // If namespace was clicked, then open it or close, it depends on the previous state.
+                if (selectedClass == catVM)
+                {
+                    catVM.IsExpanded = !catVM.IsExpanded;
+                    // Accept situation, when root category was clicked.
+                    // Don't handle it, wpf will close it by default and unselect classes.
+                    if (selectedClass is RootNodeCategoryViewModel)
+                        e.Handled = false;
+                    else
+                        e.Handled = true;
+                    break;
+                }
+
+                // If class button was clicked.
+                // Ensure, that this class is not part of current category.
+                // Then collapse category.
+                var categoryClasses = catVM.Items.FirstOrDefault(x => x is ClassesNodeCategoryViewModel);
+                if (catVM != null && categoryClasses != null)
+                    if (!(categoryClasses as ClassesNodeCategoryViewModel).Items.Contains(selectedClass)
+                        && selectedClass != catVM)
+                        catVM.IsExpanded = false;
+            }
```

Comment:
Hi @aosyatnik, can this be made in a more generic method that does not have too many preconditions? When an expansion needs to happen, I should be able to call just `ExpandClassElement` given a `NodeCategoryViewModel`, then that method will internally handle the case to see if this belongs to a given category, if yes, retain the category expansion state, otherwise collapse the category. Does that make sense?